Fix Hue emulation not working for switches

challs opened the pull request:
## Description:

Previously, the Hue emulation would only call something an on/off light
if it was a light and was not dimmable. The problem with this is that
switches, for example, are not dimmable and create confusion if exposed
as a dimmable light.

So the test is changed to only show something as dimmable if it is
a dimmable light, or it is one of the types of entities that we
map the brightness to something else (climate control, fans, media
players or covers)

Thanks to @ts8789 (#27767) / @schiegg for the idea to whitelist only
those domains for which the brightness is interpreted.

This PR differs from #27767 as follows:

- Script entities are left as dimmable instead of converting to on/off. This is because the code supports sending the dimming value to the script and therefore would be a breaking change for existing installs.
- In order to make the code slightly more maintainable (DRY), the items which do not change according to the brightness are combined together into a single assignment
- There is unit test that checks that the switch type is now an on/off switch

**Related issue (if applicable):** fixes #26860

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html

challs committed:
Fix Hue emulation not working for switches #26860

Previously, the Hue emulation would only call something an on/off light
if it was a light and was not dimmable. The problem with this is that
switches, for example, are not dimmable and create confusion if exposed
as a dimmable light.

So the test is changed to only show something as dimmable if it is
a dimmable light, or it is one of the types of entities that we
map the brightness to something else (climate control, fans, media
players or covers)

Thanks to ts8789 (PR#27767) / schiegg for the idea to whitelist only
those domains for which the brightness is interpreted.

jbcodemonkey commented:
@challs Just one comment on the change: currently the modelid for dimmable vs on/off switches is different. Dimmables have returned modelid HASS123 while on/off have returned modelid HASS321.

This change has them all returning modelid HASS321.

That may not matter at all, but it is a change in how these devices will be seen vs how they have been presented in the past.

challs commented:
> This change has them all returning modelid HASS321.
> 
> That may not matter at all, but it is a change in how these devices will be seen vs how they have been presented in the past.

Eeek, sorry I missed that. Thanks for noticing @jbcodemonkey. I've reverted back to the original behaviour. I don't really want to start making changes to anything other than fixing the type at this stage in a bid to stabilise existing setups.

challs committed:
Remove duplicate code from entity state generation function

No functional changes, purely a refactor to remove the duplicate
lines and make the differences between the dimmable and on/off
types clearer

challs commented:
I think my work has been superseeded by @Noblekangaroo in #28920 so I'm going to close this one.
